sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,525 @@
-      break;
-    case kGpioIrqTriggerEdgeFallingLevelHigh:
-      mmio_region_write32(gpio->params.base_addr, GPIO_INTRPT_RISE_EN_REG_OFFSET, 1 << pin);
-      mmio_region_write32(gpio->params.base_addr, GPIO_INTRPT_FALL_EN_REG_OFFSET, 1 << pin);
-      mmio_region_write32(gpio->params.base_addr, GPIO_INTRPT_LVL_LOW_EN_REG_OFFSET, 1 << pin);
-      break;
-    default:
-      return kGpioError;
-  }
-
-  return kGpioOk;
+gpio_result_t gpio_en_input_sampling (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    gpio_peri->GPIO_EN0 = bitfield_write(gpio_peri->GPIO_EN0,
+                                         MASK_1b, pin, GPIO_EN__ENABLED);
+    return GpioOk;
 }
 
-gpio_result_t gpio_read(const gpio_t *gpio, gpio_pin_t pin,
-                                bool *state) {
-  if (gpio == NULL || state == NULL) {
-    return kGpioBadArg;
-  }
+gpio_result_t gpio_dis_input_sampling (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    gpio_peri->GPIO_EN0 = bitfield_write(gpio_peri->GPIO_EN0,
+                                         MASK_1b, pin, GPIO_EN__DISABLED);
+    return GpioOk;
+}
 
-  *state = mmio_region_get_bit32(gpio->params.base_addr,
-                                 GPIO_GPIO_IN_REG_OFFSET, pin);
+gpio_result_t gpio_reset (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+
+    gpio_intr_set_mode (0);
+    gpio_set_mode (pin, GpioModeIn);
+    gpio_dis_input_sampling (pin);
+    gpio_peri->GPIO_CLEAR0 = bitfield_write(gpio_peri->GPIO_CLEAR0,
+                                         MASK_1b, pin, GPIO_CLEAR__REMOVE_MASK);
+    gpio_peri->GPIO_SET0 = bitfield_write(gpio_peri->GPIO_SET0,
+                                         MASK_1b, pin, GPIO_SET__REMOVE_MASK);
+    gpio_intr_dis_all(pin);
+    gpio_intr_clear_stat(pin);
+    return GpioOk;
+}
 
-  return kGpioOk;
+gpio_result_t gpio_reset_all (void)
+{
+    gpio_peri->GPIO_MODE0 = 0;
+    gpio_peri->GPIO_MODE1 = 0;
+    gpio_peri->GPIO_EN0 = 0;
+    gpio_peri->GPIO_CLEAR0 = 0;
+    gpio_peri->GPIO_SET0 = 0;
+    gpio_peri->GPIO_TOGGLE0 = 0;
+    gpio_peri->INTRPT_RISE_EN0 = 0;
+    gpio_peri->INTRPT_FALL_EN0 = 0;
+    gpio_peri->INTRPT_LVL_HIGH_EN0 = 0;
+    gpio_peri->INTRPT_LVL_LOW_EN0 = 0;
+    gpio_peri->INTRPT_STATUS0 = 0xFFFFFFFF;
 }
 
-gpio_result_t gpio_write(const gpio_t *gpio, gpio_pin_t pin,
-                                 bool state) {
+gpio_result_t gpio_read (gpio_pin_number_t pin, bool *val)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    if (bitfield_read(gpio_peri->GPIO_IN0, MASK_1b, pin) == GPIO_IS_SET)
+        *val = true;
+    else
+        *val = false;
+    return GpioOk;
+}
 
-  if (gpio == NULL) {
-    return kGpioBadArg;
-  }
+//todo: check it
+gpio_result_t gpio_toggle (gpio_pin_number_t pin)
+{
```

Comment:
why are you not using the `GPIO_TOGGLE0` register?

Reply:
I used but it was not working properly so I decided to do it by checking and toggling. However, probably it was my fault, if its better to be implemented in that way, I will check it.

Reply:
ok